modules/gdscript/gdscript_format.h

```diff
@@ -0,0 +1,201 @@
+/*************************************************************************/
+/*  gdscript_format.h                                                    */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef GDSCRIPT_FORMAT_H
+#define GDSCRIPT_FORMAT_H
+
+#include "core/string/ustring.h"
+#include "gdscript_parser.h"
+
+using GDP = GDScriptParser;
+
+class GDScriptFormat {
+private:
+	enum BreakType {
+		NONE,
+		WRAP,
+		SPLIT
+	};
+
+	typedef String (GDScriptFormat::*ParserFunc)(GDP::Node *, int, BreakType);
+	HashSet<int> new_lines;
+
+private:
+	static bool node_has_comments(const GDP::Node *p_node);
+	static bool children_have_comments(const GDP::Node *p_parent);
+
+	static int get_operation_priority(const GDP::BinaryOpNode::OpType p_op_type) {
```

Comment:
look at `GDScriptParser::get_rule(token)`

I believe a BinaryOpNode knows what precedence value is via `get_rule(current.type)->precedence`

That's just based on looking in the code just now, so it might not work out.  But it is something you probably need to explain why the formatter has its own opinion about the language definition, if you can't make it work with the data that is in the tree you are working with.

Maybe it works though then you could dramatically simplify.

Reply:
get_rule() operates off of GDScriptTokenizer's Token's Type enum, which does not match BinaryOpNode's Type list, and there is no mechanic to reverse engineer the token from which a Node came from.

So ultimately, using get_rule would:
- need to store the parser in a class variable
- need a function to translate the type of binary operator optype into a token type
- get the rule from the parser using the translation

So I feel it'd more complicated to set up the architecture there than it is just having a precedence rule function for this one node.

Reply:
I agree.  I had searched poorly.  The code was referring to the token 'current' that is only transient during parsing, and the node never stores the precedence that it dereferences there.   That's too bad.   It still feels to me the facility you created is weird to exist outside the language definition.  If it is in fact possible to infer precedence JUST From a node type, then it should be possible to just have a table that maps from node types to token types and then look up the token in the parser.  That way, if the language changes you would get updates.   If you mapped multiple tokens, you could even check if there is any contradiction in your code.  Like if the language actually does NOT strictly sort those BinaryOpNode types because of context.  Just some ideas you might or might not find interesting.